### PR TITLE
Chore: Change so we cache loading plugins by its version

### LINKED ISSUE #16771: [FeatureRequest] Plugins - Better cache-busting on plugin assets
**What would you like to be added**:
I would like to run the server with an environment variable/flag containing an ID to be used for cache busting, rather than using `Date.now()` on the client as seen in https://github.com/grafana/grafana/blob/7aeae84c523fda5f694758d2eb6aac6053809c81/public/app/features/plugins/plugin_loader.ts#L35-L39

**Why is this needed**:
The current implementation ensures that nothing is ever in the cache, while in reality we could use a content hash/git commit hash/etc. to do much better. This needlessly slows down grafana (in our case by hundreds of milliseconds as our users of our Grafana setup are spread out across US and Europe).

I can provide a patch if you give some pointers as to how you'd like to see this implemented.

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
This is one of the follow up PRs from the escalation described in this one: https://github.com/grafana/grafana/pull/41347

We have changed so we will cache plugins (javascript files) based on their version number instead of the `Date.now()` that previously was used.

We have decided to still keep the `Date.now()` cache buster for the plugins css and text imports. So these won't be cached. The reasoning behind this is to see the results of improving the caching of the javascript in the first step so see if it is worth applying the same fix for css and text which is a bit tricker.

**Todo:**
- [x] Introduce caching by plugin version number.
- [x] Invalidate cache when updating/removing plugins via the catalog
- [x] Add tests for the cache buster.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/support-escalations/issues/1167
Fixes #16771

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by marefr** [comment]
Nice. Something to keep in mind is that maybe _caching by plugin version number_ should only apply when `application_mode = production` to allow a smooth experience when developing plugins, e.g. not having to clear caches between changes etc.

File: public/app/features/plugins/admin/state/actions.ts [hunk 1]
```diff
@@ -63,6 +64,9 @@ export const install = createAsyncThunk(
     try {
       await installPlugin(id, version);
       await updatePanels();
+      if (isUpdating) {
```
Comment (hugohaggmark) [inline public/app/features/plugins/admin/state/actions.ts:67]: Any tests that need to change/update because of this change?

**Comment by hugohaggmark** [comment]
super nit: I think if you move this line https://github.com/grafana/grafana/blob/e9d90231e0c30ec250405c516abd0a4091444eb2/.github/CODEOWNERS#L66 between https://github.com/grafana/grafana/blob/e9d90231e0c30ec250405c516abd0a4091444eb2/.github/CODEOWNERS#L92-L93 user-essentials wont be pinged about plugin changes in the future.

File: public/app/features/plugins/admin/state/actions.ts (hunk 1 above)
Comment (mckn) [reply public/app/features/plugins/admin/state/actions.ts:67]: I think so. Let me have a look.

**Comment by mckn** [comment]
> Nice. Something to keep in mind is that maybe _caching by plugin version number_ should only apply when `application_mode = production` to allow a smooth experience when developing plugins, e.g. not having to clear caches between changes etc.

It looks like we are setting the headers to `no-cache` in development. So this should not be a problem 🤞🏻

File: public/app/app.ts [hunk 2]
```diff
@@ -119,8 +119,8 @@ export class GrafanaApp {
 
       // Preload selected app plugins
       const promises: Array<Promise<any>> = [];
-      for (const modulePath of config.pluginsToPreload) {
-        promises.push(importPluginModule(modulePath));
+      for (const plugin of config.pluginsToPreload) {
+        promises.push(importPluginModule(plugin.path, plugin.version));
       }
```
Comment (torkelo) [inline public/app/app.ts:124]: Can we take this opportunity and move this out of the main app boot so a failed plugin load does not cause all of Grafana to crash unrecoverably

Comment (mckn) [reply public/app/app.ts:124]: Yes! We were planning on doing this in a separate PR. Will link it shortly.

Comment (mckn) [reply public/app/app.ts:124]: @torkelo https://github.com/grafana/grafana/pull/41490

Comment (torkelo) [reply public/app/app.ts:124]: :+1:

**Review by dprokop — APPROVED** [review]
LGTM @mckn!

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**